### ISSUE TITLE
html features: preserve whitespaces

### ISSUE THREAD

**Issue by kmike** [TeamHG-Memex/eli5#31]
Features with whitespaces in front get these whitespaces removed in HTML.

Compare:

```
+2.837  spa 
+2.805   spa
```

and 

![2016-10-21 16 59 04](https://cloud.githubusercontent.com/assets/107893/19602989/b773245e-97af-11e6-8367-758de2b418fa.png)

I think whitespaces should be replaced with `&nbsp;` for HTML display. It could also make sense to use another background for text, in order to show whitespaces in the end.

**Comment by lopuhin** [comment]
I tried some approaches for making whitespace at the end visible: changing background, adding border, replacing it with single and double underscore (https://en.wikipedia.org/wiki/Underscore), white and black square, etc., but at the end solution I like the most is to just put features having space at the start or at the end in double quotes. It's also possible to apply it to just features having space at the end (as a space at the start is visible by indentation level), but for some reason applying this rule to space both at the start and at the end feels more natural. Here is how it looks like:
![2016-10-21 19 34 37](https://cloud.githubusercontent.com/assets/424613/19605800/f3f9a0b8-97c5-11e6-8f57-9ca9b79e0b28.png)

**Comment by kmike** [comment]
The main problem with `"` is that it can be a part of a feature name, so feature names become ambiguous. Are there unambiguous options which look slightly worse, or do they all look much worse?

**Comment by lopuhin** [comment]
This was my second best, a double underscore, I almost went with it before thinking of quoting:
![2016-10-21 19 23 38](https://cloud.githubusercontent.com/assets/424613/19606294/ff0b98c2-97c9-11e6-8409-0547e3f58345.png)
But it's also not strictly unambiguous (we could also just use more fancy quotes).

Let me think of something reasonably looking and really unambiguous.

**Comment by lopuhin** [comment]
Thanks for the inspiration @kmike ! In the editor context I liked dots as spaces better, but in our case I think just highlighting spaces looks better.
I use just two colors here: darker green for positive and darker red for negative. I tried adjusting highlighting color according to the current weight color (making a darker version of weight color instead of using a constant color), but it looked worse.
Here is a notebook with new space highlighting scheme: https://github.com/TeamHG-Memex/eli5/blob/visible-space/notebooks/explain_text_prediction_char.ipynb

**Comment by kmike** [comment]
Yeah, I agree highlighting looks better.
What do you think about making it less bright, e.g. using S=70% and L=70%?
![2016-10-24 14 26 17](https://cloud.githubusercontent.com/assets/107893/19645623/f1ebb964-99f5-11e6-8ed5-fe1805e65e97.png)

If the letter which is close to a whitespace is l or other similar letter it is a bit hard to distinguish with dark highlighting:

![2016-10-24 14 27 31](https://cloud.githubusercontent.com/assets/107893/19645655/20653900-99f6-11e6-989f-93ceee35cebc.png)

**Comment by kmike** [comment]
What about using light shade instead of medium shade for whitespaces?
![2016-10-24 14 32 29](https://cloud.githubusercontent.com/assets/107893/19645779/bf2481cc-99f6-11e6-9388-30c1a845be9b.png)

![2016-10-24 14 33 11](https://cloud.githubusercontent.com/assets/107893/19645786/cf8882a2-99f6-11e6-9bd4-2f10c6978da1.png)

**Comment by kmike** [comment]
dots are nice, but sometimes they can look cryptic:

![2016-10-24 14 38 08](https://cloud.githubusercontent.com/assets/107893/19645943/8bdd00a4-99f7-11e6-8ac3-d5ae1c9b7c99.png)

**Comment by lopuhin** [comment]
Yeah, light shade looks much better for text!

About the darkness of color for html: I would rather fix it by adding a little margin, because the current darkness is barely visible over red of max intensity (there is already additional margin here but not looking like I would like):
![2016-10-24 15 42 20](https://cloud.githubusercontent.com/assets/424613/19646069/7d18bc6c-9a00-11e6-9e3e-d80a3a8413e4.png)

Maybe lowering the max intensity of red and green could help, I'll check that.

**Comment by lopuhin** [comment]
Agreed, dots look nice with letters but no so nice with other dots :)

**Comment by kmike** [comment]
The margin looks good, I like it. Re max. density: text also is not very visible on the darkest red, maybe we should find a lighter color.

**Comment by lopuhin** [comment]
I raised the minimal lightness, here are the new colors: https://github.com/TeamHG-Memex/eli5/blob/visible-space/notebooks/explain_text_prediction_char.ipynb

**Comment by kmike** [comment]
Hm, now it became harder to see highlighting in text: for example, is this text classified as positive or negative, and what are important features?

![2016-10-24 15 12 27](https://cloud.githubusercontent.com/assets/107893/19646995/5bc56852-99fc-11e6-911f-046efc1696b5.png)

(it is positive)

**Comment by lopuhin** [comment]
Right, it became too pale. I partly corrected that problem 	6604b2c by using different color schemes for weights in table and text, but I think this particular example is also an indicator of another problem - the longer the text, the paler will the in-text features be compared to the BIAS, which stays the same.
